Framework/Algorithms/src/JoinRuns.cpp

```diff
@@ -0,0 +1,401 @@
+      *first, g_log, sampleLogsSum, sampleLogsTimeSeries, sampleLogsList,
+      sampleLogsWarn, sampleLogsWarnTolerances, sampleLogsFail,
+      sampleLogsFailTolerances);
+
+  auto it = m_inputWS.begin();
+
+  // Temporary workspace to carry the merged sample logs
+  // This is cloned from the first workspace and does not have
+  // the correct size to be the output, since the size is unknown
+  // at this point. We can check only later which ones are going
+  // to be skipped, to compute the size of the output respectively.
+  MatrixWorkspace_uptr temp = first->clone();
+
+  // First sequentially merge the sample logs and build the x-axis
+  for (++it; it != m_inputWS.end(); ++it) {
+    // attempt to merge the sample logs
+    try {
+      sampleLogsBehaviour.mergeSampleLogs(**it, *temp);
+      sampleLogsBehaviour.setUpdatedSampleLogs(*temp);
+      std::vector<double> axisIt = getXAxis(*it);
+      xAxis.insert(xAxis.end(), axisIt.begin(), axisIt.end());
+    } catch (std::invalid_argument &e) {
+      if (sampleLogsFailBehaviour == SKIP_BEHAVIOUR) {
+        g_log.error() << "Could not join workspace: " << (*it)->getName()
+                      << ". Reason: \"" << e.what() << "\". Skipping.\n";
+        sampleLogsBehaviour.resetSampleLogs(*temp);
+        // remove the skipped one from the list
+        m_inputWS.erase(it);
+        --it;
+      } else {
+        throw std::invalid_argument(e);
+      }
+    }
+  }
+
+  if (m_inputWS.size() == 1) {
+    g_log.warning() << "Nothing left to join [after skipping the workspaces "
+                       "that failed to merge the sample logs].";
+    // note, we need to continue still, since
+    // the x-axis might need to be changed
+  }
+
+  // now get the size of the output
+  size_t outBlockSize = xAxis.size();
+  size_t numSpec = first->getNumberHistograms();
+
+  m_outWS = WorkspaceFactory::Instance().create(first, numSpec, outBlockSize,
+                                                outBlockSize);
+
+  // copy over the merged sample logs from the temp
+  m_outWS->mutableRun() = temp->run();
+
+  m_progress = make_unique<Progress>(this, 0.0, 1.0, numSpec);
+
+  // Now loop in parallel over all the spectra and join the data
+  PARALLEL_FOR_IF(threadSafe(*m_outWS))
+  for (long index = 0; index < static_cast<long>(numSpec); ++index) {
+    PARALLEL_START_INTERUPT_REGION
+    m_outWS->mutableX(static_cast<size_t>(index)) = xAxis;
+    joinSpectrum(index);
+    m_progress->report();
+    PARALLEL_END_INTERUPT_REGION
+  }
+  PARALLEL_CHECK_INTERUPT_REGION
+
+  if (!m_logEntry.empty()) {
+    std::string unit = first->run().getLogData(m_logEntry)->units();
+    try {
+      m_outWS->getAxis(0)->unit() = UnitFactory::Instance().create(unit);
+    } catch (Exception::NotFoundError &) {
+      m_outWS->getAxis(0)->unit() = UnitFactory::Instance().create("Empty");
+    }
+  }
+
+  setProperty("OutputWorkspace", m_outWS);
+}
+} // namespace Algorithms
+} // namespace Mantid
```

Comment:
I think there needs to be some tests that at least touch the sample log behaviour to check they are wired up correctly. It does not need as many as in `MergeRuns`, but at least something that touches the input options.